### PR TITLE
[17.0][MIG]sale_order_line_input: Migration to version 17.0

### PR DESCRIPTION
**Pull request by juanjosesegui-tecnativa**
@Tecnativa TT50121

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please check CI. Remember to pass pre-commit always before submitting. It can be trigger automatically if you do `pre-commit install -f` in the repository.

**Comment by pedrobaeza** [comment]
@sbidoul I see the problem comes from this commit that entered the other day:

https://github.com/OCA/OCB/commit/85b2ede4ee59bf18cdbee4ef97e9da3f023887c2

The DB should have `pg_trgm` extension installed.

**Comment by dreispt** [comment]
@sbidoul @pedrobaeza It seems it has been fixed in the meanwhile: https://github.com/odoo/odoo/pull/173101
I'm shocked at how lightly an additional system dependency was introduced, that would break most Odoo installations out there...

**Comment by pedrobaeza** [comment]
Yes, it has been tracked in https://github.com/OCA/oca-ci/issues/74, but they have found and solved soon the problem. It was not their intention to have the extra dependency when optimizing it. Anyway, it's very interesting to have the extension on production databases to benefit from these optimizations.

File: sale_order_line_input/views/sale_order_line_view.xml [hunk 1]
```diff
@@ -157,32 +157,30 @@
...
                     name="product_packaging_qty"
-                    attrs="{'invisible': ['|', ('product_id', '=', False), ('product_packaging_id', '=', False)]}"
+                    column_invisible="'|', product_id == False, product_packaging_id == False"
                     groups="product.group_stock_packaging"
                     optional="show"
                 />
                 <field
                     name="product_packaging_id"
-                    attrs="{'invisible': [('product_id', '=', False)]}"
+                    column_invisible="product_id == False"
```
Comment (chienandalu) [inline sale_order_line_input/views/sale_order_line_view.xml]: It's probably valid in the views, but better state it as:
```suggestion
column_invisible="not product_id"
```

File: sale_order_line_input/views/sale_order_line_view.xml [hunk 2]
```diff
@@ -123,32 +120,35 @@
...
                 <field
                     name="product_uom"
                     force_save="1"
                     string="UoM"
-                    attrs="{
-                        'readonly': [('product_uom_readonly', '=', True)],
-                        'required': [('display_type', '=', False)],
-                    }"
+                    readonly="product_uom_readonly"
+                    required="display_type == False"
```
Comment (chienandalu) [inline sale_order_line_input/views/sale_order_line_view.xml]: same here. we want to coherce any falsey value not just `False`:
```
                    required="not display_type"
```

File: sale_order_line_input/views/sale_order_line_view.xml [hunk 3]
```diff
@@ -78,11 +77,9 @@
...
                     name="product_template_id"
                     string="Product"
-                    invisible="1"
-                    attrs="{
-                        'readonly': [('product_updatable', '=', False)],
-                        'required': [('display_type', '=', False)],
-                    }"
+                    column_invisible="1"
+                    readonly="product_updatable == False"
+                    required="display_type == False"
```
Comment (chienandalu) [inline sale_order_line_input/views/sale_order_line_view.xml]:
```
                    readonly="not product_updatable"
                    required="not display_type"
```

File: sale_order_line_input/views/sale_order_line_view.xml [hunk 4]
```diff
@@ -47,18 +48,16 @@
...
+                <field name="product_type" column_invisible="1" />
+                <field name="product_updatable" column_invisible="1" />
                 <field
                     name="product_id"
-                    attrs="{
-                        'readonly': [('product_updatable', '=', False)],
-                        'required': [('display_type', '=', False)],
-                    }"
+                    readonly="product_updatable == False"
+                    required="display_type == False"
```
Comment (chienandalu) [inline sale_order_line_input/views/sale_order_line_view.xml]:
```
                    readonly="not product_updatable"
                    required="not display_type"
```

File: sale_order_line_input/views/sale_order_line_view.xml [hunk 5]
```diff
@@ -47,18 +48,16 @@
                     context="{'res_partner_search_mode': 'customer', 'default_is_company': True, 'default_customer_rank': 1}"
                     options='{"always_reload": True}'
                     required="1"
-                    attrs="{'readonly': [('order_partner_id', '!=', False), ('order_id', '!=', False)]}"
+                    readonly="'order_id', 'order_partner_id'"
```
Comment (chienandalu) [inline sale_order_line_input/views/sale_order_line_view.xml]: You have to do a python statement equivalent to the former attr readonly domain.

File: sale_discount_display_amount/README.rst
Comment (chienandalu) [inline sale_discount_display_amount/README.rst:1]: Looks like pre-commit brought changes to another module in the repo. It would be prefereble to have them in another commit at least (or even in another repo!)

**Review by chienandalu — CHANGES_REQUESTED** [review]
Also rebase to fix the CI

**Comment by pedrobaeza** [comment]
What is missing here?

**Comment by juanjosesegui-tecnativa** [comment]
> What is missing here?

codecov/patch was returning an error for the _onchange_order_partner_id function because it was not fully covered by the test. I have modified it, and it is now correct.

**Review by sanderlienaerts — COMMENTED** [review]
An error occurred during functional tests. 

1. Open the Sales Order Lines list view
2. Create a new sale order line
3. Select a product first

After selecting the product, the error below occurs. ~~Seems related to `product_supplierinfo_for_customer_sale`. Can you confirm?~~ Disabling the module does not fix the issue. 

```
Traceback (most recent call last):
  File "/opt/odoo/odoo/models.py", line 5856, in ensure_one
    _id, = self._ids
ValueError: not enough values to unpack (expected 1, got 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/odoo/odoo/http.py", line 1764, in _serve_db
    return service_model.retrying(self._serve_ir_http, self.env)
  File "/opt/odoo/odoo/service/model.py", line 133, in retrying
    result = func()
  File "/opt/odoo/odoo/http.py", line 1791, in _serve_ir_http
    response = self.dispatcher.dispatch(rule.endpoint, args)
  File "/opt/odoo/odoo/http.py", line 1995, in dispatch
    result = self.request.registry['ir.http']._dispatch(endpoint)
  File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 222, in _dispatch
    result = endpoint(**request.params)
  File "/opt/odoo/odoo/http.py", line 741, in route_wrapper
    result = endpoint(self, *args, **params_ok)
  File "/opt/odoo/addons/web/controllers/dataset.py", line 24, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/addons/web/controllers/dataset.py", line 20, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 468, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 453, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/opt/odoo/addons/web/models/models.py", line 1073, in onchange
    todo = [
  File "/opt/odoo/addons/web/models/models.py", line 1076, in <listcomp>
    if field_name not in done and snapshot0.has_changed(field_name)
  File "/opt/odoo/addons/web/models/models.py", line 1189, in has_changed
    return self[field_name] != self.record[field_name]
  File "/opt/odoo/odoo/models.py", line 6638, in __getitem__
    return self._fields[key].__get__(self, self.env.registry[self._name])
  File "/opt/odoo/odoo/fields.py", line 1152, in __get__
    self.recompute(record)
  File "/opt/odoo/odoo/fields.py", line 1367, in recompute
    apply_except_missing(self.compute_value, recs)
  File "/opt/odoo/odoo/fields.py", line 1340, in apply_except_missing
    func(records)
  File "/opt/odoo/odoo/fields.py", line 1389, in compute_value
    records._compute_field_value(self)
  File "/opt/odoo/odoo/models.py", line 4882, in _compute_field_value
    fields.determine(field.compute, self)
  File "/opt/odoo/odoo/fields.py", line 102, in determine
    return needle(*args)
  File "/opt/odoo/addons/sale_management/models/sale_order_line.py", line 16, in _compute_name
    super()._compute_name()
  File "/mnt/data/odoo-addons-dir/product_supplierinfo_for_customer_sale/models/sale_order_line.py", line 42, in _compute_name
    super(SaleOrderLine, item)._compute_name()
  File "/opt/odoo/addons/sale/models/sale_order_line.py", line 319, in _compute_name
    lang = line.order_id._get_lang()
  File "/opt/odoo/addons/sale/models/sale_order.py", line 1948, in _get_lang
    self.ensure_one()
  File "/opt/odoo/odoo/models.py", line 5859, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: sale.order()

The above server error caused the following client error:
RPC_ERROR: Odoo Server Error
    RPCError@http://oca-sale-workflow-17-0-pr3232-b8d138ccacb4.runboat.odoo-community.org/web/assets/302b793/web.assets_web.min.js:2883:324
    makeErrorFromResponse@http://oca-sale-workflow-17-0-pr3232-b8d138ccacb4.runboat.odoo-community.org/web/assets/302b793/web.assets_web.min.js:2886:175
    @http://oca-sale-workflow-17-0-pr3232-b8d138ccacb4.runboat.odoo-community.org/web/assets/302b793/web.assets_web.min.js:2890:34

```

**Comment by pedrobaeza** [comment]
@sanderlienaerts the problem seems to be with module `product_supplierinfo_for_customer_sale`. Uninstall it for your tests.

File: sale_order_line_input/views/sale_order_line_view.xml [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+                    optional="show"
+                />
+                <field
+                    name="customer_lead"
+                    optional="hide"
+                    readonly="state not in ('draft', 'sent', 'sale')"
+                />
+                <field
+                    name="product_packaging_qty"
+                    invisible="'|', not product_id, not product_packaging_id"
```
Comment (carlos-lopez-tecnativa) [inline sale_order_line_input/views/sale_order_line_view.xml]: Please check this expression is wrong, convert to python expression
```suggestion
                    invisible="not product_id or not product_packaging_id"
```

File: sale_order_line_input/views/sale_order_line_view.xml [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+                    type="object"
+                    icon="fa-external-link"
+                    class="oe_stat_button"
+                />
+                <field
+                    name="order_partner_id"
+                    context="{'res_partner_search_mode': 'customer', 'default_is_company': True, 'default_customer_rank': 1}"
+                    options='{"always_reload": True}'
+                    required="1"
+                    readonly="('order_partner_id', '!=', False), ('order_id', '!=', False)"
```
Comment (carlos-lopez-tecnativa) [inline sale_order_line_input/views/sale_order_line_view.xml]: check expression

File: sale_order_line_input/views/sale_order_line_view.xml [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+                        'company_id': company_id
+                    }"
+                />
+                <field
+                    name="qty_delivered"
+                    decoration-info="(not display_type and invoice_status == 'to invoice')"
+                    decoration-bf="(not display_type and invoice_status == 'to invoice')"
+                    string="Delivered"
+                    invisible="state != 'sale'"
+                    readonly="'qty_delivered_method' != 'manual'"
```
Comment (carlos-lopez-tecnativa) [inline sale_order_line_input/views/sale_order_line_view.xml]: I think it must be evaluated as an expression instead of a string.
```suggestion
                    readonly="qty_delivered_method != 'manual'"
```

File: sale_order_line_input/views/sale_order_line_view.xml [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+                <field name="is_downpayment" column_invisible="1" />
+                <field
+                    name="price_subtotal"
+                    widget="monetary"
+                    invisible="is_downpayment"
+                />
+                <field
+                    name="price_total"
+                    widget="monetary"
+                    invisible="(is_downpayment),('tax_calculation_rounding_method', '=' ,'round_globally')"
```
Comment (carlos-lopez-tecnativa) [inline sale_order_line_input/views/sale_order_line_view.xml]: and here

**Review by carlos-lopez-tecnativa — CHANGES_REQUESTED** [review]
@juanjosesegui-tecnativa as a suggestion, you could use [this module](https://github.com/OCA/server-tools/tree/17.0/views_migration_17) to facilitate the migration of attrs to Python expressions

**Comment by dreispt** [comment]
> @juanjosesegui-tecnativa as a suggestion, you could use [this module](https://github.com/OCA/server-tools/tree/17.0/views_migration_17) to facilitate the migration of attrs to Python expressions

That work has potential to be converted into a great contribution to https://github.com/OCA/odoo-module-migrator
cc @legalsylvain

File: sale_order_line_input/views/sale_order_line_view.xml [hunk 10]
```diff
@@ -9,29 +10,30 @@
...
         <field name="arch" type="xml">
             <tree
-                decoration-info="invoice_status=='to invoice'"
-                decoration-warning="state=='draft'"
-                decoration-muted="state=='cancel'"
-                decoration-success="state=='done'"
+                decoration-info="invoice_status =='to invoice'"
+                decoration-warning="state =='draft'"
+                decoration-muted="state =='cancel'"
+                decoration-success="order_locked == True"
```
Comment (chienandalu) [inline sale_order_line_input/views/sale_order_line_view.xml]: I'd simply change this to decoration-success="invoice_status =='invoiced'" and get rid of the `order_locked` stuff

Comment (pilarvargas-tecnativa) [reply sale_order_line_input/views/sale_order_line_view.xml]: Yes, because the locked field of sale.order is a flag that is controlled by the setting of blocking confirmed sales orders, so it is not valid to use it as a flag. Thanks

**Comment by pilarvargas-tecnativa** [comment]
> An error occurred during functional tests.
> 
> 1. Open the Sales Order Lines list view
> 2. Create a new sale order line
> 3. Select a product first
> 
> After selecting the product, the error below occurs. ~Seems related to `product_supplierinfo_for_customer_sale`. Can you confirm?~ Disabling the module does not fix the issue.

You can try again, it should be solved. Thanks

File: sale_order_line_input/views/sale_order_line_view.xml (hunk 10 above)
Comment (pilarvargas-tecnativa) [reply sale_order_line_input/views/sale_order_line_view.xml]: I have added the changes and also reverted the pre-commit changes in another module.

**Comment by pilarvargas-tecnativa** [comment]
@carlos-lopez-tecnativa Can you check again? Thanks

File: sale_order_general_discount/README.rst
Comment (chienandalu) [inline sale_order_general_discount/README.rst:1]: Some changes from other module slipped into your PR

Comment (pilarvargas-tecnativa) [reply sale_order_general_discount/README.rst:1]: removed

**Review by chienandalu — APPROVED** [review]
Test and code :+1:

**Review by carlos-lopez-tecnativa — APPROVED** [review]
LGTM thanks
